### PR TITLE
ZK-3907: session timeout message ignores application wide preferred l…

### PR DESCRIPTION
**Pull request by klyvechen**
…ocale

### CODE REVIEW

File: zktest/test/java/org/zkoss/zktest/zats/test2/B85_ZK_3907Test.java [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+
+import static org.zkoss.web.servlet.Charsets.*;
+
+
+public class B85_ZK_3907Test {
+	
+	@Test
+	public void checkGetPreferredLocale() {
+		Library.setProperty(Attributes.PREFERRED_LOCALE, "es_PY");
+		Locale l = getPreferredLocale(null, new ServletRequest() {
```
Comment (scribetw) [inline zktest/test/java/org/zkoss/zktest/zats/test2/B85_ZK_3907Test.java]: Just use mock library (easymock etc) to mock a `ServletRequest` interface and do what you want.

File: zweb/src/org/zkoss/web/servlet/Charsets.java [hunk 2]
```diff
@@ -208,13 +210,11 @@ public static final Locale getPreferredLocale(HttpSession sess, ServletRequest r
 					return (Locale) v;
 				logLocaleError(v);
 			}
-
-			final String s = Library.getProperty(Attributes.PREFERRED_LOCALE);
-			if (s != null)
-				return Locales.getLocale(s);
+
```
Comment (scribetw) [inline zweb/src/org/zkoss/web/servlet/Charsets.java]: Clear the tab

File: zweb/src/org/zkoss/web/servlet/Charsets.java [hunk 3]
```diff
@@ -190,6 +190,8 @@ public static final void markSetup(ServletRequest request, boolean setup) {
 	 * @param sess the session to look for the preferred locale. Ignored if null.
 	 */
 	public static final Locale getPreferredLocale(HttpSession sess, ServletRequest request) {
+		final String s = Library.getProperty(Attributes.PREFERRED_LOCALE);
+		Locale l = (s != null) ? Locales.getLocale(s) : request.getLocale();
```
Comment (scribetw) [inline zweb/src/org/zkoss/web/servlet/Charsets.java]: `final Locale l` for safety

File: zktest/test/java/org/zkoss/zktest/zats/test2/B85_ZK_3907Test.java (hunk 1 above)
Comment (klyvechen) [reply zktest/test/java/org/zkoss/zktest/zats/test2/B85_ZK_3907Test.java]: ok

File: zweb/src/org/zkoss/web/servlet/Charsets.java (hunk 2 above)
Comment (klyvechen) [reply zweb/src/org/zkoss/web/servlet/Charsets.java]: ok

File: zweb/src/org/zkoss/web/servlet/Charsets.java (hunk 3 above)
Comment (klyvechen) [reply zweb/src/org/zkoss/web/servlet/Charsets.java]: ok

**Pull request merged**